traefik 1.4.0 (new formula)

vonagam opened the pull request:
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,37 @@
+
+    cd dir do
+      system "go", "generate"
+      system "go", "build", "-o", bin/"traefik", "./cmd/traefik"
```
prefix.install_metafiles

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,37 @@
+  end
+
+  test do
+    system bin/"traefik", "version"
```
The test will need to exercise and verify the functionality of the software once it has been installed, not just run the version command.

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,37 @@
+
+class Traefik < Formula
+  desc "Træfik, a modern reverse proxy"
+  homepage "https://traefik.io"
```
https://traefik.io/

vonagam commented:
Ok, have a question. I have found out that build with go is incomplete, because webgui is not get bundled in, because it does not get built.

There are two options: use other provided build way - docker, or to just grab executable from releases page https://github.com/containous/traefik/releases/tag/v1.3.8

Second way seems to me easier to implement and is faster to install, but there are two builds for darwin: 386 and amd64. So, should i create two formulas Traefik64 and Traefik32?

First way seems to have one advantage though, it will allow to install head.

Which way is more preferable?

ilovezfs commented:
In Homebrew we have to build from source, so just shipping the upstream binaries isn't an option. We also cannot build with Docker on our CI. So we'll either need to go with the webgui, or figure out how to build it from source outside of Docker. Probably the former.

ilovezfs commented:
@BrewTestBot test this please

vonagam commented:
I will return to it on this week. Will add requested changes and add dependency on yarn to build web dashboard.

ilovezfs commented:
@BrewTestBot test this please

vonagam commented:
Implemented requested changes (slash at the end of homepage url, `prefix.install_metafiles`, test). Added building of web dashboard. Bumped version to last stable (1.3.8).

vonagam commented:
@ilovezfs are any additional changes needed?

vonagam commented:
@ilovezfs @fxcoudert ping.

ilovezfs commented:
@BrewTestBot test this please

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,61 @@
+class Traefik < Formula
+  desc "Træfik, a modern reverse proxy"
+  homepage "https://traefik.io/"
+  url "https://github.com/containous/traefik/archive/v1.3.8.tar.gz"
```
This should use the release tarball https://github.com/containous/traefik/releases/download/v1.3.8/traefik-v1.3.8.src.tar.gz

vonagam replied:
Should i comment here instead of root level to communicate completion of requested changes?

vonagam commented:
@ilovezfs now uses release tarball.

vonagam commented:
@ilovezfs @fxcoudert any update?

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,61 @@
+  end
+
+  test do
+    (testpath/"traefik.toml").write <<-EOF.undent
```
should use `<<~EOS` not `<<-EOF.undent`

vonagam replied:
Ok.
( I just did what other formulas have: there are "<<-EOF.undent", but no "<<~EOS" )

ilovezfs replied:
Check the formulas now. Nothing using `.undent` anymore ;)

vonagam replied:
Ok, at the time of pull request creation...

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,61 @@
+  def install
+    ENV["GOPATH"] = buildpath
+    dir = buildpath/"src/github.com/containous/traefik"
+    dir.install buildpath.children
```
`(buildpath/"src/github.com/containous/traefik").install buildpath.children` is less than 80 character so we don't need the intermediate variable

vonagam replied:
But this variable used below too.
And this is a usual occurrence in formulas. (search "dir = buildpath/")

ilovezfs replied:
The general rule is not to introduce `dir.install` unless it goes over 80 here.

vonagam replied:
But what about variable reuse?

ilovezfs replied:
As a general rule of life: fewer lines = fewer bugs.

vonagam replied:
Done.

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,61 @@
+class Traefik < Formula
+  desc "Træfik, a modern reverse proxy"
+  homepage "https://traefik.io/"
+  url "https://github.com/containous/traefik/releases/download/v1.3.8/traefik-v1.3.8.src.tar.gz"
```
needs an explicit
```
  version "1.3.8"
```
because the version detected from that url is `1.3.8.src` not `1.3.8`

vonagam replied:
Done.

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,61 @@
+        exec bin/"traefik", "--configfile=#{testpath}/traefik.toml"
+      end
+      sleep 5
+      output = shell_output("curl -sIm3 -XGET http://localhost:45678/dashboard/")
```
instead of hard-coding a port, get an open one with
```
    require "socket"

    server = TCPServer.new(0)
    port = server.addr[1]
    server.close
```
and then use port

vonagam replied:
Ok. Done, i think.

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,71 @@
+    server.close
+
+    server = TCPServer.new(0)
+    http_port = server.addr[1]
```
you could potentially end up with the same one twice since you closed the first one already so get both before closing either

vonagam replied:
Right catch, fixed.

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,70 @@
+        exec bin/"traefik", "--configfile=#{testpath}/traefik.toml"
+      end
+      sleep 5
+      output = shell_output("curl -sIm3 -XGET http://localhost:#{web_port}/dashboard/")
```
use
```
      cmd = "curl -sIm3 -XGET http://localhost:#{web_port}/dashboard/"
      assert_match /200 OK/m, shell_output(cmd)
```
to avoid going passed the eightieth column

vonagam replied:
Done.

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,70 @@
+require "language/go"
+
+class Traefik < Formula
+  desc "Træfik, a modern reverse proxy"
```
There's an audit rule against starting the description with the formula name, which you've managed to circumvent, but still applies :)

vonagam replied:
Heh)
Well, that's what they have in description of github repository.
What should i write?
Just "A modern reverse proxy"?

vonagam replied:
Pushed with that.
By the way, where can i find those audit rules?

ilovezfs replied:
https://github.com/Homebrew/brew/blob/master/Library/Homebrew/dev-cmd/audit.rb
https://github.com/Homebrew/brew/tree/master/Library/Homebrew/rubocops

in general just

brew audit --new-formula traefik

ilovezfs commented on Formula/traefik.rb:
```diff
@@ -0,0 +1,70 @@
+require "language/go"
+
+class Traefik < Formula
+  desc "A modern reverse proxy"
```
There's an audit rule against starting the description with an indefinite article ("a" or "an")

vonagam replied:
Done.
Same question as before, are those audit rules are published somewhere?

ilovezfs commented:
This looks good to me, if you're happy.

vonagam commented:
Ok, thanks.

ilovezfs commented:
Thanks for your first contribution to Homebrew, @vonagam! Without people like you submitting PRs we couldn't run this project. You rock!
